libs/sync/diff.go

```diff
@@ -15,6 +19,71 @@ func (d diff) IsEmpty() bool {
 	return len(d.put) == 0 && len(d.delete) == 0
 }
 
+// Compute operations required to make files in WSFS reflect current local files.
+// Takes into account changes since the last sync iteration.
+func computeDiff(target *SnapshotState, current *SnapshotState) diff {
+	d := &diff{
+		delete: make([]string, 0),
+		rmdir:  make([]string, 0),
+		mkdir:  make([]string, 0),
+		put:    make([]string, 0),
+	}
+	d.addRemovedFiles(target, current)
+	d.addFilesWithRemoteNameChanged(target, current)
+	d.addNewFiles(target, current)
+	d.addUpdatedFiles(target, current)
+	return *d
+}
+
+// Add operators for tracked files that no longer exist.
+func (d *diff) addRemovedFiles(after *SnapshotState, before *SnapshotState) {
+	for localName, remoteName := range before.LocalToRemoteNames {
+		if _, ok := after.LocalToRemoteNames[localName]; !ok {
+			d.delete = append(d.delete, remoteName)
+		}
+	}
+
+	// Remove directories that would no longer contain any files.
+	beforeDirs := MakeDirSet(maps.Keys(before.LocalToRemoteNames))
+	afterDirs := MakeDirSet(maps.Keys(after.LocalToRemoteNames))
+	d.rmdir = append(d.rmdir, beforeDirs.Remove(afterDirs).Slice()...)
```

Comment:
I don't think we actually ever append to this. Can we use assignment?

Reply:
done